Sortable table

breeku opened the pull request:
Addresses https://github.com/psikoi/wise-old-man/issues/136

![https://i.imgur.com/aYdICkT.gif](https://i.imgur.com/aYdICkT.gif)

Sorting modes are default, ascending and descending. Clicking loops through them

- [x] Functionality

- [x] CSS

psikoi commented:
Oh nice! Join the discord man, we talk dev in there

psikoi commented on app/src/components/Table/Table.scss:
```diff
@@ -44,11 +44,51 @@
       font-size: 0;
       padding: 0;
     }
+
```
Can you use css modifiers for this?

```
.arrow {
   
    &.-ascending {
       *override the default values here*
    }
   
    &.-descending{
       *override the default values here*
    }


    *default arrow attributes here*
}
```

psikoi commented on app/src/components/Table/Table.jsx:
```diff
@@ -23,17 +69,26 @@ function Table({ rows, columns, highlightedIndex, onRowClicked, clickable }) {
-              <th className={className && className()} key={`col-${key}`}>
+              <th className={customClass} key={`col-${key}`} onClick={() => handleClick(key)}>
                 {label || label === '' ? label : capitalize(key)}
+                <div className={`${customClass} arrow`}>
```
Following my comment below about modifiers, you could then use the `classNames` package imported at the top of the file to do something like:

`
className = {"arrow", {"-ascending": isAscending, "-descending": isDescending}`

psikoi commented on app/src/components/Table/Table.jsx:
```diff
@@ -1,15 +1,61 @@
+    if (sortData.type === SORT.ASCENDING) {
+      return b[sortData.by] - a[sortData.by];
+    }
+    if (sortData.type === SORT.DESCENDING) {
```
Mind adding a line above that second "if"?

psikoi commented on app/src/components/Table/Table.jsx:
```diff
@@ -1,15 +1,61 @@
+    return 0;
+  };
+
+  useEffect(() => {
```
What is this for? Is it for restoring the default sorting when the table gets unmounted?

breeku replied:
Yeah, otherwise if you sort by example by 'rank' in skilling, then change the tab to bossing, 'rank' will be sorted by default. Not sure if that's a bad thing though.

psikoi replied:
ah makes sense

psikoi commented:
I renamed some variables and added isSortable prop to table columns (there are some columns where sorting isn't needed).

However there's a bigger issue with this, the sorting only works for direct columns, for example:

```
{ 
    key: 'rank', 
    width: 70 
},
{
  key: 'start',
  transform: val => <NumberLabel value={val} />,
  className: () => '-break-large',
  get: row => (row.progress ? row.progress.start : 0)
}
```

It will work for `rank`, since the object key is "row", but for `start `for example, it has a custom "get" method, and those columns aren't being sorted, as row[start] is undefined (it should be row.progress.start).

I'll do some refactor on the Table component in general since it's getting pretty messy now, once it's cleaner, i'll fix this PR to correctly sort all columns 👍

psikoi commented:
- Refactored Table component
- Fixed string sorting (it wasn't sorting string values like "username")

breeku commented:
> 
> 
> I renamed some variables and added isSortable prop to table columns (there are some columns where sorting isn't needed).
> 
> However there's a bigger issue with this, the sorting only works for direct columns, for example:
> 
> ```
> { 
>     key: 'rank', 
>     width: 70 
> },
> {
>   key: 'start',
>   transform: val => <NumberLabel value={val} />,
>   className: () => '-break-large',
>   get: row => (row.progress ? row.progress.start : 0)
> }
> ```
> 
> It will work for `rank`, since the object key is "row", but for `start `for example, it has a custom "get" method, and those columns aren't being sorted, as row[start] is undefined (it should be row.progress.start).
> 
> I'll do some refactor on the Table component in general since it's getting pretty messy now, once it's cleaner, i'll fix this PR to correctly sort all columns 👍

That's a good catch, I somehow didn't even try the table on other than the player page. My bad...

> 
> 
>     * Refactored Table component
> 
>     * Fixed custom column "get" selector sorting
> 
>     * Fixed string sorting (it wasn't sorting string values like "username")
> 
> 
> @breeku take a look at the code and let me know if I'm missing something 👍

Did you notice these warnings
`Line 57:6:  React Hook useEffect has a missing dependency: 'onTimerEnded'. Either include it or remove the dependency array  react-hooks/exhaustive-deps` 
and
`Line 75:64:  React Hook useMemo has a missing dependency: 'handleSort'. Either include it or remove the dependency array  react-hooks/exhaustive-deps`

Otherwise looks good! From what I tested now it works well :)
